Controller/ExportController.php

```diff
@@ -91,6 +92,13 @@ function () use (
                                 'ignore_paginator' => 1,
                             ]
                         );
+
+                        $utmValues = $this->getUtmTagsByLeadIds($batch);
+                        $utmFields = $this->getFieldNames($utmValues, ['lead_id']);
+
+                        $deviceValues = $this->getDeviceDataByLeadIds($batch);
+                        $deviceFields = $this->getFieldNames($deviceValues, ['lead_id']);
```

Comment:
One issue...

What if the first lead in a batch has no utm data or no device data? 
We're generating a CSV by batch, so the column count must be constant throughout, assuming we find all data for all leads.

Similarly, the anon function needs to always add the same column quantity... 
In this kind of situation, it may be better to hard-code the column keys to ensure consistency.